common/internal_model/src/main/scala/uk/ac/wellcome/models/parse/Marc008DateParser.scala

```diff
@@ -0,0 +1,77 @@
+package uk.ac.wellcome.models.parse
+
+import fastparse._, NoWhitespace._
+
+import uk.ac.wellcome.models.work.internal.InstantRange
+import DateParserImplicits._
+
+/**
+  *  Parses Marc 008 date information into InstantRange
+  *
+  *  Spec: https://www.loc.gov/marc/bibliographic/bd008a.html
+  */
+object Marc008DateParser extends Parser[InstantRange] with DateParserUtils {
+
+  def parser[_: P] =
+    singleKnownDate |
+      multipleDates.toInstantRange |
+      detailedDate.toInstantRange |
+      publicationDateAndCopyrightDate |
+      reprintDate |
+      continuingResourceCeasedPublication.toInstantRange |
+      continuingResourceCurrentlyPublished.toInstantRange |
+      continuingResourceStatusUnknown.toInstantRange |
+      questionableDate.toInstantRange |
+      differingReleaseAndProduction.toInstantRange
+
+  def singleKnownDate[_: P] =
+    ("s" ~ partialYear ~ emptyDate)
+
+  def multipleDates[_: P] =
+    ("m" ~ year ~ year) map { case (from, to) => FuzzyDateRange(from, to) }
+
+  def detailedDate[_: P] =
+    ("e" ~ yearDigits ~ digitRep(exactly = 2) ~ digitRep(exactly = 2))
+      .map { case (year, month, day) => CalendarDate(day, month, year) }
+
+  def publicationDateAndCopyrightDate[_: P] =
+    ("t" ~ partialYear ~ partialYear)
+      .map { case (pubYear, copyYear) => pubYear }
+
+  def reprintDate[_: P] =
+    ("r" ~ partialYear ~ partialYear)
+      .map { case (reprintYear, originalYear) => reprintYear }
+
+  def continuingResourceCeasedPublication[_: P] =
+    ("d" ~ year ~ year) map { case (from, to) => FuzzyDateRange(from, to) }
+
+  // TODO : InstantRange should have optional start / end rather than 9999
+  def continuingResourceCurrentlyPublished[_: P] =
+    ("c" ~ year ~ "9999") map (FuzzyDateRange(_, Year(9999)))
+
+  // TODO : InstantRange should have optional start / end rather than 9999
+  def continuingResourceStatusUnknown[_: P] =
+    ("u" ~ year ~ "uuuu") map (FuzzyDateRange(_, Year(9999)))
+
+  // TODO : should these dates be marked as inferred?
+  def questionableDate[_: P] =
+    ("q" ~ year ~ year) map { case (from, to) => FuzzyDateRange(from, to) }
+
+  def differingReleaseAndProduction[_: P] =
+    ("p" ~ year ~ year) map { case (release, production) => release }
+
+  // TODO : should century / century and decade be marked inferred
+  def partialYear[_: P] =
```

Comment:
Is there a reason we use `.toInstantRange` on these parsers, and not do this when setting `val parser`?

Just made a bit tricky to read and understand why [some parsers were being converted, and other had already been converted](https://github.com/wellcometrust/catalogue/pull/80/files#diff-340b42773e0489ec2e33b6d0ea6499eeR17-R20) - but not that confusing.

Reply:
If we don't convert these parsers here, i.e we have

```scala
def partialYear[_: P] = year | century | centuryAndDecade
```

then the return type of this method is `P[FuzzyDate]` rather than `P[InstantRange]`.

There is no implicit `ToInstantRange[FuzzyDate]` (as some `FuzzyDate` types are not specific unless to convert to `InstantRange`) so we wouldn't then be able to call the `toInstantRange` method on the result of this.

It is slightly annoying, but is required to do it this way if we want the strong typing of what can be converted to an `InstantRange` (otherwise we would have to return `Option[InstantRange]` like we were doing previously).

Reply:
Yea - it was just a readability thing, I think keeping the strong type system makes more sense though - my 2¢.